cmd/{build,up/local}: add option to set linker flags

darkowlzz opened the pull request:
<!--

Before making a PR, please read our contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD

Note: Make sure your branch is rebased to the latest upstream master.

-->

**Description of the change:**
This adds two new flags to `build` and `up local` commands:
`shell` and `ldflags`.

`ldflags` flag is used to pass the go build -ldflag options to the
underlying operator build go command.

`shell` flag is used to set the shell which runs the go build command.
This is needed because just executing `go` using `exec.Command` results
in some argument parsing issue, resulting in incorrect command syntax,
with error:
```
flag provided but not defined: -ldflags "-X import/path/foo.version
usage: go build [-o output] [-i] [build flags] [packages]
Run 'go help build' for details.
```
Symbols like `=` in the flag limits the parsing and using escape
characters don't work in that context. Running go via a shell works.
Hence, this adds a flag to set the shell. It defaults to "bash". It can
be replaced by other shells as well. Bash, zsh and csh support `-c` flag
when passing an executable command.


**Motivation for the change:**
#730 

Closes #730 

<!--

Note: If this PR is fixing an issue make sure to add a note saying:

-->

estroz commented:
@darkowlzz once #563 is merged, the `go build` step will be removed and all binary builds will be in [Docker containers](https://github.com/operator-framework/operator-sdk/pull/563/files#diff-a9d989728aa9fdaa951d8e0cfd4ebb07L148), so we don't need the `operator-sdk build` flags.

estroz commented on commands/operator-sdk/cmd/up/local.go:
```diff
@@ -100,12 +104,13 @@ func mustKubeConfig() {
 func upLocal() {
-	args := []string{"run", filepath.Join(scaffold.ManagerDir, scaffold.CmdFile)}
+	linkerOpts := fmt.Sprintf("-ldflags \"%s\"", ldFlags)
+	args := []string{"go", "run", linkerOpts, filepath.Join(scaffold.ManagerDir, scaffold.CmdFile)}
```
You're seeing expansion issues because `args` in `exec.Command(name, args)` are split by list element and not spaces, ex. `exec.Command("go", []string{"run", "-ldflags \"-X version.Version=1.0.0\""}...)` is not the same as `exec.Command("go", []string{"run", "-ldflags", "-X version.Version=1.0.0"}...)`.

The following works for me:
```Go
func upLocal() {
    	args := []string{"run"}
	if ldFlags != "" {
		args = append(args, []string{"-ldflags", ldFlags}...)
	}
	args = append(args, filepath.Join(scaffold.ManagerDir, scaffold.CmdFile))
	if operatorFlags != "" {
		args = append(args, strings.Split(operatorFlags, " ")...)
	}
	dc := exec.Command("go", args...)
        ....
}
```
The `--shell` flag is therefore not needed.

estroz commented on commands/operator-sdk/cmd/build.go:
```diff
@@ -60,6 +62,8 @@ For example:
 	buildCmd.Flags().StringVar(&testLocationBuild, "test-location", "./test/e2e", "Location of tests")
 	buildCmd.Flags().StringVar(&namespacedManBuild, "namespaced-manifest", "deploy/operator.yaml", "Path of namespaced resources manifest for tests")
+	buildCmd.Flags().StringVar(&shell, "shell", "bash", "The shell where the build command is run; default to bash")
+	buildCmd.Flags().StringVar(&ldFlags, "ldflags", "", "Set Go linker options")
```
`--go-ldflags` is more informative.

darkowlzz commented:
@estroz thanks a lot. That did work.
I've updated the PR and removed build command changes. `--go-ldflags` flag is added for `up local` command only.
